### PR TITLE
Merge Data Source

### PR DESCRIPTION
**Pull request by igoroctaviano**
### Context

A new data source that allows merging one or more data sources at a series level.
Implementation can be expended later to allow the merge at the study level.

### Changes & Results

- Add new merge data source
- Attach new merge data source to default extension

### Testing

```
window.config = {
  ...
  dataSources: [
    {
      sourceName: 'merge',
      namespace: '@ohif/extension-default.dataSourcesModule.merge',
      configuration: {
        name: 'merge',
        friendlyName: 'Merge dicomweb-1 and dicomweb-2 data at series level',
        seriesMerge: {
          dataSourceNames: ['dicomweb-1', 'dicomweb-2'],
        },
      },
    },
    {
      sourceName: 'dicomweb-1',
      ...
    },
    {
      sourceName: 'dicomweb-2',
      ...
    },
  ],
};

```

#### PR

- [] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [] My code has been well-documented (function documentation, inline comments,
  etc.)

#### Public Documentation Updates

- [] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [] OS: 
- [] Node version: 
- [] Browser:
  

[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch

### CODE REVIEW

File: platform/core/src/services/DicomMetadataStore/DicomMetadataStore.ts
Comment (sedghi) [inline platform/core/src/services/DicomMetadataStore/DicomMetadataStore.ts:193]: what about if there is no series with that UID

File: extensions/default/src/MergeDataSource/index.js
Comment (sedghi) [inline extensions/default/src/MergeDataSource/index.js:58]: should this be `data = data.concat(incomingData)`?

File: extensions/default/src/MergeDataSource/utils.js
Comment (sedghi) [inline extensions/default/src/MergeDataSource/utils.js:40]: && obj[currentKey] !== null

File: extensions/default/src/MergeDataSource/utils.js
Comment (sedghi) [inline extensions/default/src/MergeDataSource/utils.js:39]: how about for of

File: extensions/default/src/MergeDataSource/index.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/index.js]: Updated, thanks!

File: platform/core/src/services/DicomMetadataStore/DicomMetadataStore.ts
Comment (igoroctaviano) [reply platform/core/src/services/DicomMetadataStore/DicomMetadataStore.ts:198]: Updated, thanks!

File: extensions/default/src/MergeDataSource/utils.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/utils.js]: Updated, thanks!

File: extensions/default/src/MergeDataSource/utils.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/utils.js]: Updated, thanks!

**Review by sedghi — COMMENTED** [review]
looks good see my comments

File: extensions/default/src/DicomWebDataSource/retrieveStudyMetadata.js
Comment (wayfarer3130) [inline extensions/default/src/DicomWebDataSource/retrieveStudyMetadata.js:42]: In general making it look URI ish is nice => how about `${dicomWebConfig.name}:${StudyInstanceUID}`

File: extensions/default/src/MergeDataSource/index.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/index.js:82]: globalData can be const since you add to it but don't reassign it
Also, an easy way to key it for uniqueness is
const globalData = new Map<...>()
then set the values by whatever key they use, and at the end, 
return [...globalData.values()]

Also, can we call is mergedData rather than globalData?

File: extensions/default/src/MergeDataSource/index.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/index.js:91]: You need to promises.push(promise.then(data => ....
Otherwise you have introduced a very subtle bug when the allSettled can be complete, but the merge is not, and the length ends up changing after the response is returned.

File: extensions/default/src/MergeDataSource/index.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/index.js:122]: This should be === as I think the includes sourceName is tricky to ensure you get it right.  
If you want a way to group calls, then add a parameter to the configuration that allows you to compare for exact equality on that parameter value.  You could make that parameter keyed so that someone could create a new/arbitrary key for this.

File: extensions/default/src/MergeDataSource/index.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/index.js:138]: This is NOT the default - the default is specifically specified by name as a parameter 'defaultDataSourceName' or some such value.  Also, the default might be specific to the merge configuration, so I think you should have two levels of default to allow the merge configuration to specify a specific default.

File: extensions/default/src/MergeDataSource/index.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/index.js:155]: Suggest adding a getter to get your default data source name based on the current circumstances.

Comment (wayfarer3130) [reply extensions/default/src/MergeDataSource/index.js:155]: Or a const that you compute up front.

File: extensions/default/src/MergeDataSource/utils.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/utils.js:8]: Consider if you can use the lodash get method for this.

File: extensions/default/src/MergeDataSource/utils.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/utils.js:38]: The keys aren't hierarchical, they are a flat denormalized structure, so return obj[key] is sufficient.

**Review by wayfarer3130 — CHANGES_REQUESTED** [review]
A number of smaller changes, but this looks like it is on the right track.

File: extensions/default/src/MergeDataSource/index.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/index.js:122]: The includes here is running against an array, and in that case, it's the same as ===. You can test it by running ['xpto'].includes('xpt') which will return false.

File: extensions/default/src/DicomWebDataSource/retrieveStudyMetadata.js
Comment (igoroctaviano) [reply extensions/default/src/DicomWebDataSource/retrieveStudyMetadata.js]: Updated, thanks!

File: extensions/default/src/MergeDataSource/index.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/index.js]: Updated, thanks!

File: extensions/default/src/MergeDataSource/index.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/index.js]: Updated, thanks!

File: extensions/default/src/MergeDataSource/index.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/index.js]: Updated, thanks!

File: extensions/default/src/MergeDataSource/utils.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/utils.js]: Updated, thanks!

File: extensions/default/src/MergeDataSource/utils.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/utils.js]: Updated, thanks!

File: platform/docs/docs/platform/extensions/modules/data-source.md
Comment (sedghi) [inline platform/docs/docs/platform/extensions/modules/data-source.md:173]: Love this thanks

File: platform/docs/docs/platform/extensions/modules/data-source.md
Comment (sedghi) [inline platform/docs/docs/platform/extensions/modules/data-source.md:177]: this is ambigious wording, do you mean whoever arrives first?

**Review by sedghi — APPROVED** [review]

File: platform/docs/docs/platform/extensions/modules/data-source.md
Comment (igoroctaviano) [reply platform/docs/docs/platform/extensions/modules/data-source.md]: Updated wording, thanks!

File: extensions/default/src/MergeDataSource/index.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/index.js:1]: Could you move this to typescript - it would make figuring out a few of the calls a bit easier, as I wasn't always sure of the types.

File: extensions/default/src/MergeDataSource/index.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/index.js:55]: The merge map can be made global/static as nothing in it is specific to the function itself.

File: extensions/default/src/MergeDataSource/index.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/index.js:14]: It is quite a bit more efficient to do
mergedData.has(get(r,mergeKey)) instead of keys.includes, which performs a linear search.  On 2000 item series, that is 2,000,000 comparisons - definitely something to avoid if possible.

Comment (wayfarer3130) [reply extensions/default/src/MergeDataSource/index.js:14]: What you end up doing is const key = get(r,mergeKey);  mergedData.set(key, r)  You can then also add it to a return list, OR you can choose which version to include based on other conditions (future work)

File: extensions/default/src/MergeDataSource/index.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/index.js:30]: You should not be calling flat here - the UID keys are always at the top level.  Anything below that is extraneous and might result in missing data if a sub-sequence has an item in it.  Also makes it REALLY slow because in DICOM JSON you will have 1000+ lists in each instance, so you can end up comparing tens of thousands of instances in a linear compare.

File: extensions/default/src/MergeDataSource/index.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/index.js:175]: This one will only be for the specified data source.

File: extensions/default/src/MergeDataSource/index.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/index.js:177]: The metadata will be for a single data source unless you merge multiple same series into one.

File: extensions/default/src/MergeDataSource/index.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/index.js:61]: This one had better be 'query.series.search': {....
Once you have merged the series, then you don't merge metadata instances.

File: extensions/default/src/MergeDataSource/index.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/index.js:14]: I'm not sure about has, this is not part of array prototype, can you sketch some code sample?

File: extensions/default/src/MergeDataSource/index.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/index.js:30]: Can you suggest where to execute flat then? this call is called once per series. I'm not sure why this will be called for every instance.

File: extensions/default/src/MergeDataSource/index.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/index.js]: Updated, thanks.

File: extensions/default/src/MergeDataSource/index.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/index.js]: Updated, thanks.

File: extensions/default/src/MergeDataSource/index.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/index.js]: Updated, thanks.

File: extensions/default/src/MergeDataSource/index.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/index.js]: This method checks for retrieveAETitle and filters out the other data sources, in the end is going to be one dataSource. This is necessary since we don't know which data source to call beforehand.

File: extensions/default/src/MergeDataSource/index.js
Comment (igoroctaviano) [reply extensions/default/src/MergeDataSource/index.js]: Updated, thanks.

File: extensions/default/src/MergeDataSource/index.test.js
Comment (wayfarer3130) [inline extensions/default/src/MergeDataSource/index.test.js:1]: Thank you for adding tests for this - that will be important as we add more functionality

**Review by sedghi — APPROVED** [review]

**Review by wayfarer3130 — APPROVED** [review]
LGTM - this looks like a solid base to build more advanced handling of different types.

**Pull request merged**
